RFD 4: Release Process (#6661)
gravitational/teleport: rfd/0004-release-process.md

---
authors: Russell Jones (rjones@goteleport.com)
state: draft
---

# RFD 4 - Release Plan

## What/Why

While the Teleport team has grown, the release process has remained ad-hoc with varying release managers following varying steps with varying outcomes.

> This RFD proposes a formalized process allows spreading the load of managing a release and de-silo knowledge and lead to a smoother release process.

Reviewer: Feels like the word "that" is missing between "process" and "allows".

## Details

### Release Manager Selection

The release manager will be rotated across team leads (or a delegated proxy) for every release.

The current list of rotation release managers is below.

* @awly for Security.
* @russjones for Scale.
* @russjones for SSH and Kubernetes Access.
* @r0mant for Application and Database Access.
* @russjones for Release.

### Release Manager Responsibilities

The release manager is responsible for shepherding the release before, during, and after a release in the areas outlined below.

> Release managers will track their process by forking `.github/release.md` for each release and including it in the milestone. A sample `release.md` is included below and will act as a dashboard to view status of the release.

Reviewer 2: ```suggestion
Release managers will track their process by creating a GitHub issue using the "Release" template for each release and including it in the milestone. A sample issue template is included below and will act as a dashboard to view status of the release.
```

> #### Track Approvals

Reviewer 3: @russjones It will help to have some fallback/redundancy for both steps

Release managers will synchronize with and wait for approval of the following individuals responsible for the outlined areas before cutting a release.

* @awly for Security.
* @russjones for Scale.
* @russjones for SSH and Kubernetes Acccess.
* @r0mant for Application and Database Access.
* @russjones for Release.
* @klizhentas for Documentation.
* @klizhentas for Product.

#### Track Release Date

Release managers will synchronize with the approvers list above and update the release date accordingly.

> To update the release date, release managers need approval from @klizhentas, after which the release manger will:

Reviewer: manger -> manager

1. Submit a PR to update the [Upcoming Releases](https://goteleport.com/docs/preview/upcoming-releases/) to notify customers.

> 2. Send an email to Engineering, Customer Success, and Sales about the updated release date so they can inform customers or prospects waiting for a release appropriately.

Author: @stevenGravy Does your team have a alias it would like to be notified at?

> 2. Send an email to Engineering, Customer Success, and Sales about the updated release date so they can inform customers or prospects waiting for a release appropriately.

Author: @travelton Does your team have an alias it would like to be notified at?

Reviewer 4: Yes! customer-success@goteleport.com would be awesome.

> 2. Send an email to Engineering, Customer Success, and Sales about the updated release date so they can inform customers or prospects waiting for a release appropriately.

Author: We should notify marketing as well.

#### Test Plan

> 
> Release managers will decide which parts of the test plan are appropriate for each release and round-robin assign them to individual team members.

Reviewer 2: ```suggestion

Release managers will create a separate GitHub issue from the "Release test plan" template.
Release managers will decide which parts of the test plan are appropriate for each release and round-robin assign them to individual team members.

Major releases should pass the full test plan.
Minor releases can skip the sections that were not changed since the previous release.
```

In general, two weeks should be allocated for testing a major release and one week for minor releases.

#### Release Process

Release managers will follow the process outlined below for each release after testing is complete.

> * Prepare CHANGELOG/Release Notes. Must be approved be @klizhentas.

Reviewer 2: ```suggestion
* Prepare CHANGELOG/Release Notes. Must be approved by @klizhentas.
```

> * Block merging into release branches (needs investigation).

Reviewer 2: what's this about?

* Create release branch.
* TBD.
* Cut release.
* Organize release retrospective.

> ----

Reviewer 2: ```suggestion
----
** Release plan issue template **
```

> # Teleport x.y.z Release Plan

Reviewer: You can just include this section in `.github/release.md` template and link from here?

> # Teleport x.y.z Release Plan

Reviewer 2: ```suggestion
# Teleport x.y Release Plan
```

## Release Manager

> Russell Jones

Reviewer 2: ```suggestion
Your Name @your-github-handle
```

## Release Date

- [ ] First Beta: January 1st, 1970.
- [ ] First RC/Testing Start: January 1st, 1970.
- [ ] Release: January 1st, 1970.

> ## Product Approvals

Reviewer 2: Change `Product Approvals` to `Release Approvals` and add customer success and marketing.

- [ ] Security @awly
- [ ] Scale @russjones
- [ ] SSH and Kubernetes Access @russjones

> - [ ] Application and Database Access @russjones

Reviewer 2: ```suggestion
- [ ] Application and Database Access @r0mant
```

> - [ ] Application and Database Access @russjones

Reviewer: `@r0mant`

- [ ] Release @russjones
- [ ] Documentation @klizhentas
- [ ] Product @klizhentas

## Test Plan Approval

> - [ ] Link to Teleport x.y.z Test Plan.

Reviewer 2: ```suggestion
- [ ] Link to Teleport x.y Test Plan.
```

> # Release Process

Reviewer 2: Add a step to connect customer success and marketing activities around the release.
For example, if marketing publishes a blog post, existing customers can be sent a link.
